### PR TITLE
Extension to read and ingest Delta Lake tables

### PR DESCRIPTION
**Pull request by abhishekrb19**
Delta Lake is an open source storage layer that brings reliability to data lakes. Users can ingest data stored in a Delta Lake table into Apache Druid via a new input source `delta`. This is available via the Delta Lake extension, add the `druid-deltalake-extensions` to the list of loaded extensions. The Delta input source reads the configured Delta Lake table and extracts all the underlying delta files in the table's latest snapshot. Delta Lake files are versioned Parquet format.

This is joint work with @abhishekagarwal87, @LakshSingla and @AmatyaAvadhanula.

Example of a DML query using MSQ:
```sql
REPLACE INTO "delta-employee-datasource" OVERWRITE ALL
SELECT * FROM TABLE(
  EXTERN(
    '{"type":"delta","tablePath":"s3a://your-bucket/employee-delta-table"}',
    '{"type":"json"}'
  )
) EXTEND ("timestamp" VARCHAR, "name" VARCHAR, "age" BIGINT, "salary" DOUBLE, "bonus" FLOAT)
PARTITIONED BY ALL
```

Delta `ioConfig` in native batch spec :
```json
    "ioConfig": {
      "type": "index_parallel",
      "inputSource": {
        "type": "delta",
        "tablePath": "s3a://your-bucket/employee-delta-table"
      },
    }
```

Web-console screenshots:

CleanShot 2024-01-30 at 02 45 51@2x

Sample and ingestion of a mock dataset from the cloud:
CleanShot 2024-01-30 at 02 47 49@2x

[Delta 3.1.0](https://github.com/delta-io/delta/releases/tag/v3.1.0) just released. Can look into upgrading the delta dependency as a follow-up.

#### Release note

Enhanced ingestion capabilities to support ingestion of Delta Lake tables.

This PR has:

- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [x] added documentation for new or modified features or behaviors.
- [x] a release note entry in the PR description.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [x] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [x] been tested in a test Druid cluster.

### CODE REVIEW

**Comment by abhishekagarwal87** [comment]
Is Kerberos required for the first cut?

File: docs/ingestion/input-sources.md [hunk 1]
```diff
@@ -715,6 +715,13 @@ rolled-up datasource `wikipedia_rollup` by grouping on hour, "countryName", and
  to `true` to enable a compatibility mode where the timestampSpec is ignored.
 :::
 
+The [secondary partitioning method](native-batch.md#partitionsspec) determines the requisite number of concurrent worker tasks that run in parallel to complete ingestion with the Combining input source.
+Set this value in `maxNumConcurrentSubTasks` in `tuningConfig` based on the secondary partitioning method:
+- `range` or `single_dim` partitioning: greater than or equal to 1
+- `hashed` or `dynamic` partitioning: greater than or equal to 2
+
+For more information on the `maxNumConcurrentSubTasks` field, see [Implementation considerations](native-batch.md#implementation-considerations).
```
Comment (abhishekagarwal87) [inline docs/ingestion/input-sources.md:723]: seems unrelated to the PR

File: extensions-contrib/druid-deltalake-extensions/src/main/java/org/apache/druid/delta/input/DeltaInputRow.java [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+      final byte[] arr = dataRow.getBinary(columnOrdinal);
+      final char[] charArray = new char[arr.length];
+      for (int i = 0; i < arr.length; i++) {
+        charArray[i] = (char) (arr[i] & 0xff);
+      }
+      return String.valueOf(charArray);
+    } else if (dataType instanceof DecimalType) {
+      return dataRow.getDecimal(columnOrdinal).longValue();
+    } else {
+      throw InvalidInput.exception("Unsupported data type[%s]", dataType);
```
Comment (abhishekagarwal87) [inline extensions-contrib/druid-deltalake-extensions/src/main/java/org/apache/druid/delta/input/DeltaInputRow.java:216]: we should fieldName to the exception message.

File: extensions-contrib/druid-deltalake-extensions/src/main/java/org/apache/druid/delta/input/DeltaInputRow.java [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+      return dataRow.getDouble(columnOrdinal);
+    } else if (dataType instanceof StringType) {
+      return dataRow.getString(columnOrdinal);
+    } else if (dataType instanceof BinaryType) {
+      final byte[] arr = dataRow.getBinary(columnOrdinal);
+      final char[] charArray = new char[arr.length];
+      for (int i = 0; i < arr.length; i++) {
+        charArray[i] = (char) (arr[i] & 0xff);
+      }
+      return String.valueOf(charArray);
```
Comment (abhishekagarwal87) [inline extensions-contrib/druid-deltalake-extensions/src/main/java/org/apache/druid/delta/input/DeltaInputRow.java:212]: can you describe the logic? why does a single byte need to be masked with `0xff`?

File: extensions-contrib/druid-deltalake-extensions/src/main/java/org/apache/druid/delta/input/DeltaInputSource.java [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+      @Nullable InputFormat inputFormat,
+      File temporaryDirectory
+  )
+  {
+    Configuration hadoopConf = new Configuration();
+    TableClient tableClient = DefaultTableClient.create(hadoopConf);
+    try {
+      final Row scanState;
+      final List<Row> scanRowList;
+
```
Comment (abhishekagarwal87) [inline extensions-contrib/druid-deltalake-extensions/src/main/java/org/apache/druid/delta/input/DeltaInputSource.java:105]: can you add some comments describing the implementation?

File: extensions-contrib/druid-deltalake-extensions/src/main/java/org/apache/druid/delta/input/RowSerde.java [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+import io.delta.kernel.types.StructType;
+import io.delta.kernel.types.TimestampType;
+import org.apache.druid.error.InvalidInput;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Utility class to serialize and deserialize {@link Row} object.
+ * Code borrowed from <a href="https://github.com/delta-io/delta/blob/master/kernel/examples/table-reader/src/main/java/io/delta/kernel/examples/utils/RowSerDe.java">
```
Comment (abhishekagarwal87) [inline extensions-contrib/druid-deltalake-extensions/src/main/java/org/apache/druid/delta/input/RowSerde.java:53]: can you add whats different here from the original class?

File: extensions-contrib/druid-deltalake-extensions/src/test/java/org/apache/druid/delta/input/DeltaInputSourceTest.java [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class DeltaInputSourceTest
+{
+  @Test
+  public void testReadDeltaLakeFilesSample() throws IOException
+  {
+    final DeltaInputSource deltaInputSource = new DeltaInputSource(DeltaTestUtil.DELTA_TABLE_PATH, null);
+    Assert.assertNotNull(deltaInputSource);
```
Comment (abhishekagarwal87) [inline extensions-contrib/druid-deltalake-extensions/src/test/java/org/apache/druid/delta/input/DeltaInputSourceTest.java:43]: why? :)

File: extensions-contrib/druid-deltalake-extensions/src/test/java/org/apache/druid/delta/input/DeltaInputSourceTest.java [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+
+    InputSourceReader inputSourceReader = deltaInputSource.reader(DeltaTestUtil.SCHEMA, null, null);
+    Assert.assertNotNull(inputSourceReader);
```
Comment (abhishekagarwal87) [inline extensions-contrib/druid-deltalake-extensions/src/test/java/org/apache/druid/delta/input/DeltaInputSourceTest.java:46]: inputSourceReader can never be null.

File: extensions-contrib/druid-deltalake-extensions/src/test/resources/create_delta_table.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+
+import os
+
+import argparse
+import delta
+import mimesis
+import pyspark
+
+
+def config_spark_with_delta_lake():
```
Comment (abhishekagarwal87) [inline extensions-contrib/druid-deltalake-extensions/src/test/resources/create_delta_table.py:26]: if this file is borrowed from some place, please acknowledge the source in comments.

File: docs/ingestion/input-sources.md (hunk 1 above)
Comment (abhishekrb19) [reply docs/ingestion/input-sources.md:723]: Yeah, not directly related to this extension, but I noticed that this text was nested inside the Iceberg section: [Iceberg Filter Object](https://druid.apache.org/docs/latest/ingestion/input-sources#iceberg-filter-object) and seemed irrelevant. So I moved it to the Druid Input source, where it seems more relevant.

File: extensions-contrib/druid-deltalake-extensions/src/main/java/org/apache/druid/delta/input/DeltaInputRow.java (hunk 2 above)
Comment (abhishekrb19) [reply extensions-contrib/druid-deltalake-extensions/src/main/java/org/apache/druid/delta/input/DeltaInputRow.java]: added

File: extensions-contrib/druid-deltalake-extensions/src/main/java/org/apache/druid/delta/input/DeltaInputRow.java (hunk 3 above)
Comment (abhishekrb19) [reply extensions-contrib/druid-deltalake-extensions/src/main/java/org/apache/druid/delta/input/DeltaInputRow.java]: Good catch, I was trying to convert it to an unsigned value, which is not needed in this case

File: extensions-contrib/druid-deltalake-extensions/src/main/java/org/apache/druid/delta/input/DeltaInputSource.java (hunk 4 above)
Comment (abhishekrb19) [reply extensions-contrib/druid-deltalake-extensions/src/main/java/org/apache/druid/delta/input/DeltaInputSource.java:115]: Added javadocs to this function

File: extensions-contrib/druid-deltalake-extensions/src/main/java/org/apache/druid/delta/input/RowSerde.java (hunk 5 above)
Comment (abhishekrb19) [reply extensions-contrib/druid-deltalake-extensions/src/main/java/org/apache/druid/delta/input/RowSerde.java:53]: Updated, good point

File: extensions-contrib/druid-deltalake-extensions/src/test/java/org/apache/druid/delta/input/DeltaInputSourceTest.java (hunk 6 above)
Comment (abhishekrb19) [reply extensions-contrib/druid-deltalake-extensions/src/test/java/org/apache/druid/delta/input/DeltaInputSourceTest.java]: Heh, not sure, but I removed these unnecessary assert checks

File: extensions-contrib/druid-deltalake-extensions/src/test/resources/create_delta_table.py (hunk 8 above)
Comment (abhishekrb19) [reply extensions-contrib/druid-deltalake-extensions/src/test/resources/create_delta_table.py:26]: Will do. I might change this test setup a bit to get different datatype coverage, etc.

**Review by vogievetsky — APPROVED** [review]
Web console changes look good.

File: docs/development/extensions-contrib/delta-lake.md [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+  ~ "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+  ~ KIND, either express or implied.  See the License for the
+  ~ specific language governing permissions and limitations
+  ~ under the License.
+  -->
+
+## Delta Lake Ingest extension
+
+
+Delta Lake is an open source storage layer that brings reliability to data lakes. [DeltaLakeInputSource](../../ingestion/input-sources.md#delta-lake-input-source) lets
```
Comment (vtlim) [inline docs/development/extensions-contrib/delta-lake.md:28]:
```suggestion
Delta Lake is an open source storage framework that enables building a
Lakehouse architecture with various compute engines. [DeltaLakeInputSource](../../ingestion/input-sources.md#delta-lake-input-source) lets
```
Used description from docs for something slightly less marketing sounding

File: docs/development/extensions-contrib/delta-lake.md [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+  ~
+  ~ Unless required by applicable law or agreed to in writing,
+  ~ software distributed under the License is distributed on an
+  ~ "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+  ~ KIND, either express or implied.  See the License for the
+  ~ specific language governing permissions and limitations
+  ~ under the License.
+  -->
+
+## Delta Lake Ingest extension
```
Comment (vtlim) [inline docs/development/extensions-contrib/delta-lake.md:25]:
```suggestion
## Delta Lake extension
```

File: docs/development/extensions-contrib/delta-lake.md [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+These Delta Lake files are versioned Parquet files.
+
+## Version support
+
+The Delta Lake extension uses the Delta Kernel introduced in Delta Lake 3.0.0, which is compatible with Apache Spark 3.5.x.
+Older versions are unsupported, so consider upgrading to Delta Lake 3.0.x or higher to use this extension.
+
+## Known limitations
+
+- This extension only reads from the latest Delta table snapshot as the kernel API only supports that.
```
Comment (vtlim) [inline docs/development/extensions-contrib/delta-lake.md:42]:
```suggestion
- This extension relies on the Delta Kernel API and can only read from the latest Delta table snapshot.
```

File: docs/development/extensions-contrib/delta-lake.md [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+- Column filtering isn't supported yet. So the extension will read all the columns in the configured table.
```
Comment (vtlim) [inline docs/development/extensions-contrib/delta-lake.md:43]:
```suggestion
- Column filtering isn't supported. The extension reads all columns in the configured table.
```

File: docs/ingestion/input-sources.md [hunk 13]
```diff
@@ -925,7 +932,7 @@ The following is an example of a Combining input source spec:
 ## Iceberg input source
 
 :::info
- To use the Iceberg input source, add the `druid-iceberg-extensions` extension.
+You need to include the [`druid-iceberg-extensions`](../development/extensions-contrib/iceberg.md) as an extension to use the Iceberg input source.
```
Comment (vtlim) [inline docs/ingestion/input-sources.md:935]:
```suggestion
To use the Iceberg input source, load the extension [`druid-iceberg-extensions`](../development/extensions-contrib/iceberg.md).
```

File: docs/ingestion/input-sources.md [hunk 14]
```diff
@@ -1114,11 +1121,31 @@ This input source provides the following filters: `and`, `equals`, `interval`, a
 |type|Set this value to `not`.|yes|
 |filter|The iceberg filter on which logical NOT is applied|yes|
 
+## Delta Lake input source
 
+:::info
+You need to include the [`druid-deltalake-extensions`](../development/extensions-contrib/delta-lake.md) as an extension to use the Delta Lake input source.
```
Comment (vtlim) [inline docs/ingestion/input-sources.md:1127]:
```suggestion
To use the Delta Lake input source, load the extension [`druid-deltalake-extensions`](../development/extensions-contrib/delta-lake.md).
```

**Review by vtlim — COMMENTED** [review]
Some nits but docs otherwise look good!

**Review by abhishekagarwal87 — APPROVED** [review]

**Comment by pjain1** [comment]
@abhishekrb19 thanks for this. I was just going through this and wondering if delta lake kernel api supports any partitioning by time that we can leverage during indexing. Ingestion spec generally have an interval so can that be used to prune the amount of files on delta lake that is read during indexing, with my limited understanding of delta lake kernel api it supports filtering on columns while building scan. 

I have not seen many delta lake data lakes but one I seen had a base directory which hosted the metadata in `_delta_log` folder, data was partitioned by day so had subdirectories for each day of data and those directories has hourly subdirectories. So if someone needs to run hourly indexing job will the extension end up reading all data again and again ?

**Comment by abhishekrb19** [comment]
Hi @pjain1, yes, Delta tables can be partitioned by [date](https://docs.delta.io/latest/best-practices.html#choose-the-right-partition-column). Coincidentally, Delta [3.1.0](https://github.com/delta-io/delta/releases/tag/v3.1.0) released today. In 3.1.0, the kernel supports data skipping in addition to partition pruning using filter predicates, whenever applicable. I will try to find time and add this functionality soon. In its current form, the connector will do full table scans, so I think upgrading to 3.1.0 and adding support for filters will be a good next step.

**Comment by pjain1** [comment]
@abhishekrb19 thanks

File: extensions-contrib/druid-deltalake-extensions/src/main/java/org/apache/druid/delta/input/DeltaInputSourceReader.java [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+    }
+
+    @Override
+    public boolean hasNext()
+    {
+      while (currentBatch == null || !currentBatch.hasNext()) {
+        if (!filteredColumnarBatchCloseableIterator.hasNext()) {
+          return false; // No more batches or records to read!
+        }
+        currentBatch = filteredColumnarBatchCloseableIterator.next().getRows();
```
Comment (vkorukanti) [inline extensions-contrib/druid-deltalake-extensions/src/main/java/org/apache/druid/delta/input/DeltaInputSourceReader.java:116]: one qn: do we want to check if the `currentBatch` can return `hasNext` to true? because it could contain zero rows due to the selection vector.
